src/gui/qgsexpressionpreviewwidget.cpp

```diff
@@ -0,0 +1,166 @@
+}
+
+void QgsExpressionPreviewWidget::setLayer( QgsVectorLayer *layer )
+{
+  mLayer = layer;
+  mFeatureChooserWidget->setLayer( layer );
+}
+
+void QgsExpressionPreviewWidget::setExpressionText( const QString &expression )
+{
+  mExpressionText = expression;
+  refreshPreview();
+}
+
+void QgsExpressionPreviewWidget::setCurrentFeature( const QgsFeature &feature )
+{
+  // todo: update the combo box if it has been set externaly?
+  mExpressionContext.setFeature( feature );
+  refreshPreview();
+}
+
+void QgsExpressionPreviewWidget::setGeomCalculator( const QgsDistanceArea &da )
+{
+  mDa = da;
+}
+
+void QgsExpressionPreviewWidget::setExpressionContext( const QgsExpressionContext &context )
+{
+  mExpressionContext = context;
+}
+
+void QgsExpressionPreviewWidget::refreshPreview()
+{
+  // If the string is empty the expression will still "fail" although
+  // we don't show the user an error as it will be confusing.
+  if ( mExpressionText.isEmpty() )
+  {
+    mPreviewLabel->clear();
+    mPreviewLabel->setStyleSheet( QString() );
+    setExpressionToolTip( QString() );
+    emit expressionParsed( false );
+    mExpression = QgsExpression();
+  }
+  else
+  {
+    mExpression = QgsExpression( mExpressionText );
+
+    if ( mLayer )
+    {
+      // TODO: is this OK?
+      // Only set calculator if we have layer, else use default.
+      mExpression.setGeomCalculator( &mDa );
+    }
+
+    QVariant value = mExpression.evaluate( &mExpressionContext );
+    if ( !mExpression.hasEvalError() )
+    {
+      mPreviewLabel->setText( QgsExpression::formatPreviewString( value ) );
+    }
+
+    if ( mExpression.hasParserError() || mExpression.hasEvalError() )
+    {
+      QString errorString = mExpression.parserErrorString().replace( "\n", "<br>" );
+      QString tooltip;
+      if ( mExpression.hasParserError() )
+        tooltip = QStringLiteral( "<b>%1:</b>"
+                                  "%2" ).arg( tr( "Parser Errors" ), errorString );
+      // Only show the eval error if there is no parser error.
+      if ( !mExpression.hasParserError() && mExpression.hasEvalError() )
+        tooltip += QStringLiteral( "<b>%1:</b> %2" ).arg( tr( "Eval Error" ), mExpression.evalErrorString() );
+
+      mPreviewLabel->setText( tr( "Expression is invalid <a href=""more"">(more info)</a>" ) );
+      mPreviewLabel->setStyleSheet( QStringLiteral( "color: rgba(255, 6, 10,  255);" ) );
+      setExpressionToolTip( tooltip );
+      emit expressionParsed( false );
+      setParserError( mExpression.hasParserError() );
+      setEvalError( mExpression.hasEvalError() );
+    }
+    else
+    {
+      mPreviewLabel->setStyleSheet( QString() );
+      setExpressionToolTip( QString() );
+      emit expressionParsed( true );
+      setParserError( false );
+      setEvalError( false );
+    }
+  }
+}
+
+void QgsExpressionPreviewWidget::linkActivated( const QString & )
+{
+  QgsMessageViewer *mv = new QgsMessageViewer( this );
```

Comment:
Can we have that scoped only?
This (temporarily) leaks.

```suggestion
  QgsMessageViewer mv( this );
```